Sources/AppcuesKit/Push/UNUserNotificationCenter+AutoConfig.swift

```diff
@@ -0,0 +1,180 @@
+    class WeakPushMonitoring {
+        weak var value: PushMonitoring?
+
+        init (_ wrapping: PushMonitoring) { self.value = wrapping }
+    }
+}
+
+extension UNUserNotificationCenter {
+
+    static func swizzleNotificationCenterGetDelegate() {
+        // this will swap in a new getter for UNUserNotificationCenter.delegate - giving our code a chance to hook in
+        let originalScrollViewDelegateSelector = #selector(getter: self.delegate)
+        let swizzledScrollViewDelegateSelector = #selector(appcues__getNotificationCenterDelegate)
+
+        guard let originalScrollViewMethod = class_getInstanceMethod(self, originalScrollViewDelegateSelector),
+              let swizzledScrollViewMethod = class_getInstanceMethod(self, swizzledScrollViewDelegateSelector) else {
+            return
+        }
+
+        method_exchangeImplementations(originalScrollViewMethod, swizzledScrollViewMethod)
+    }
+
+    // this is our custom getter logic for the UNUserNotificationCenter.delegate
+    @objc
+    private func appcues__getNotificationCenterDelegate() -> UNUserNotificationCenterDelegate? {
+        let delegate: UNUserNotificationCenterDelegate
+
+        var shouldSetDelegate = false
+
+        // this call looks recursive, but it is not, it is calling the swapped implementation
+        // to get the actual delegate value that has been assigned, if any - can be nil
+        if let existingDelegate = appcues__getNotificationCenterDelegate() {
+            delegate = existingDelegate
+        } else {
+            // if it is nil, then we assign our own delegate implementation so there is
+            // something hooked in to listen to scroll
+            delegate = AppcuesUNUserNotificationCenterDelegate.shared
+            shouldSetDelegate = true
+        }
+
+        Swizzler.swizzle(
+            targetInstance: delegate,
+            targetSelector: NSSelectorFromString("userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:"),
+            replacementOwner: UNUserNotificationCenter.self,
+            placeholderSelector: #selector(appcues__placeholderUserNotificationCenterDidReceive),
+            swizzleSelector: #selector(appcues__userNotificationCenterDidReceive)
+        )
+
+        Swizzler.swizzle(
+            targetInstance: delegate,
+            targetSelector: NSSelectorFromString("userNotificationCenter:willPresentNotification:withCompletionHandler:"),
+            replacementOwner: UNUserNotificationCenter.self,
+            placeholderSelector: #selector(appcues__placeholderUserNotificationCenterWillPresent),
+            swizzleSelector: #selector(appcues__userNotificationCenterWillPresent)
+        )
+
+        // If we need to set a non-nil implementation where there previously was not one,
+        // swap the swizzled getter back first, then assign, then restore the swizzled getter.
+        // This is done to avoid infinite recursion in some cases observed, where a UICollectionView,
```

Comment:
this part _may_ not be needed here, was a weird issue with UICollectionView if I recall, but also probably not hurting anything.

Reply:
Right, so if we remove `self.delegate = delegate` that's ok because the getter always ends up in the else here on each get?

```swift
if let existingDelegate = appcues__getNotificationCenterDelegate() {
    delegate = existingDelegate
} else {
    // if it is nil, then we assign our own delegate implementation so there is
    // something hooked in to listen to scroll
    delegate = AppcuesUNUserNotificationCenterDelegate.shared
}
```

Or we could keep `self.delegate = delegate` (just without the double swizzle bracket) which would mean the first time we'd hit the `else` above, but all subsequent times would hit the `if` where `existingDelegate == AppcuesUNUserNotificationCenterDelegate.shared`.

Reply:
yeah I think you can move `self.delegate = delegate` right here in the `else` block, after it is created the first time. If I recall, thats what I had done initially, then ran into the other edge case with the scrolling implementation.

Reply:
Ok, I've updated to do that 👍